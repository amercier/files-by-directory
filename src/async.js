import regeneratorRuntime from 'regenerator-runtime';

/**
 * Naive Promisify implementation.
 *
 * Implementing this instead of using NodeJS `util.Promisify` allow to maintain NodeJS 6 support.
 * @see {@link https://nodejs.org/dist/latest-v8.x/docs/api/util.html#util_util_promisify_original|util.promisify(original)}
 *
 * @param {Function} fn Function of the form: (err, callback).
 * @returns {Function} A function that returns a Promise instead.
 */
export function promisify(fn) {
  return (...args) =>
    new Promise((resolve, reject) =>
      fn(...args, (err, result) => (err ? reject(err) : resolve(result))),
    );
}

/**
 * Apply a given function to each elements of a given iterable, and yield each returned value.
 *
 * @async
 * @generator
 * @param {Iterable<T>} iterable Synchronous or asynchronous iterable.
 * @param {Function: T => U} fn Synchronous or asynchronous function.
 * @returns {AsyncIterator<U>} An asynchronous iterator that yields value returned by `fn` applied to
 * each element of `iterable`.
 */
export async function* asyncMap(iterable, fn) {
  for await (const value of iterable) {
    yield fn(value);
  }
}

/**
 * Return an asynchronous generator that yields all elements from a given iterable that pass the
 * test implemented by the provided function.
 *
 * @async
 * @generator
 * @param {Iterable<T>} iterable Synchronous or asynchronous iterable.
 * @param {Function: T => boolean} filter Synchronous or asynchronous function.
 * @returns {AsyncIterator<T>} An asynchronous iterator that yields values from `iterable` where
 * `filter` return a truthy value.
 */
export async function* asyncFilter(iterable, filter) {
  for await (const value of iterable) {
    if (await filter(value)) {
      yield value;
    }
  }
}

/**
 * Apply a given generator to each element of a given iterable, and re-yield every yielded values.
 *
 * @async
 * @generator
 * @param {Iterable<T>} iterable Synchronous or asynchronous iterable.
 * @param {Function: T => Iterable<U>} generator Synchronous or asynchronous generator function.
 * @returns {AsyncIterator<U>} An asynchronous iterator that re-yields each value yielded by `generator(iterable[i])`.
 */
export async function* asyncFlattenMap(iterable, generator) {
  for await (const value of iterable) {
    yield* generator(value);
  }
}

/**
 * Get all values from an iterable.
 *
 * @param {Iterable<T>} iterable Synchronous or asynchronous iterable.
 * @returns {Promise<T[]>} A Promise resolving to an array containing all yielded values.
 */
export async function values(iterable) {
  const array = [];
  for await (const value of iterable) {
    array.push(value);
  }
  return array;
}
