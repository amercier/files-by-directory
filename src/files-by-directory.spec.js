import '@babel/polyfill'; // Required for NodeJS < 10
import { values } from './async';
import getFilesByDirectory from './files-by-directory';
import { file1a, level1, level2, level2Files, level3, level3Files } from '../fixture';

/** @test {filesByDirectory} */
describe('filesByDirectory', () => {
  it('returns an asynchronousiterator', () => {
    const iterator = getFilesByDirectory([file1a]);
    expect(iterator.next).toBeFunction();
    expect(iterator.next()).toBeInstanceOf(Promise);
  });

  it('generates asynchronously an array containing the file when path is a file', async () => {
    expect(await values(getFilesByDirectory([file1a]))).toMatchSnapshot();
  });

  it('generates asynchronously an array of files when given path is a directory containing only files', async () => {
    expect(await values(getFilesByDirectory([level3]))).toMatchSnapshot();
  });

  it('generates one array per directory when given path is a directory containing sub-directories', async () => {
    expect(await values(getFilesByDirectory([level2]))).toMatchSnapshot();
  });

  it('omits double directory paths', async () => {
    expect(await values(getFilesByDirectory([level3, level3]))).toMatchSnapshot();
  });

  it('omits double file paths', async () => {
    expect(await values(getFilesByDirectory([file1a, file1a]))).toMatchSnapshot();
  });

  it('omits descendant directory paths', async () => {
    expect(await values(getFilesByDirectory([level2, level3]))).toMatchSnapshot();
    expect(await values(getFilesByDirectory([level3, level2]))).toMatchSnapshot();
  });

  it('omits descendant file paths', async () => {
    expect(await values(getFilesByDirectory([level3, level3Files[0]]))).toMatchSnapshot();
    expect(await values(getFilesByDirectory([level3Files[0], level3]))).toMatchSnapshot();
  });

  it('groups regular files by parent path', async () => {
    expect(await values(getFilesByDirectory(level3Files))).toMatchSnapshot();
    expect(await values(getFilesByDirectory([...level2Files, ...level3Files]))).toMatchSnapshot();
    expect(await values(getFilesByDirectory([...level3Files, ...level2Files]))).toMatchSnapshot();
  });

  describe('options', () => {
    describe('excludeSymlinks', () => {
      it('exclude symbolic links when excludeSymlinks is true', async () => {
        const options = { excludeSymlinks: true };
        expect(await values(getFilesByDirectory([level3], options))).toMatchSnapshot();
        expect(await values(getFilesByDirectory([level2], options))).toMatchSnapshot();
        expect(await values(getFilesByDirectory([level1], options))).toMatchSnapshot();
        expect(
          await values(getFilesByDirectory([level3, ...level2Files], options)),
        ).toMatchSnapshot();
      });

      it('includes symbolic links when excludeSymlinks is false', async () => {
        const options = { excludeSymlinks: false };
        expect(await values(getFilesByDirectory([level3], options))).toMatchSnapshot();
        expect(await values(getFilesByDirectory([level2], options))).toMatchSnapshot();
        expect(await values(getFilesByDirectory([level1], options))).toMatchSnapshot();
        expect(
          await values(getFilesByDirectory([level3, ...level2Files], options)),
        ).toMatchSnapshot();
      });
    });
  });
});
