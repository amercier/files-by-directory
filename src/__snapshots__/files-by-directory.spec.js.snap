// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filesByDirectory generates asynchronously an array containing the file when path is a file 1`] = `
Array [
  Array [
    "fixture/level1/file1a",
  ],
]
`;

exports[`filesByDirectory generates asynchronously an array of files when given path is a directory containing only files 1`] = `
Array [
  Array [
    "fixture/level1/level2/level3/file3a",
    "fixture/level1/level2/level3/file3b",
  ],
]
`;

exports[`filesByDirectory generates one array per directory when given path is a directory containing sub-directories 1`] = `
Array [
  Array [
    "fixture/level1/level2/level3/file3a",
    "fixture/level1/level2/level3/file3b",
  ],
  Array [
    "fixture/level1/level2/file2a",
    "fixture/level1/level2/file2b",
    "fixture/level1/level2/link-to-directory",
    "fixture/level1/level2/link-to-grand-parent-directory",
    "fixture/level1/level2/link-to-parent-directory",
  ],
]
`;

exports[`filesByDirectory groups regular files by parent path 1`] = `
Array [
  Array [
    "fixture/level1/level2/level3/file3a",
    "fixture/level1/level2/level3/file3b",
  ],
]
`;

exports[`filesByDirectory groups regular files by parent path 2`] = `
Array [
  Array [
    "fixture/level1/level2/file2a",
    "fixture/level1/level2/file2b",
    "fixture/level1/level2/link-to-directory",
    "fixture/level1/level2/link-to-grand-parent-directory",
    "fixture/level1/level2/link-to-parent-directory",
  ],
  Array [
    "fixture/level1/level2/level3/file3a",
    "fixture/level1/level2/level3/file3b",
  ],
]
`;

exports[`filesByDirectory groups regular files by parent path 3`] = `
Array [
  Array [
    "fixture/level1/level2/level3/file3a",
    "fixture/level1/level2/level3/file3b",
  ],
  Array [
    "fixture/level1/level2/file2a",
    "fixture/level1/level2/file2b",
    "fixture/level1/level2/link-to-directory",
    "fixture/level1/level2/link-to-grand-parent-directory",
    "fixture/level1/level2/link-to-parent-directory",
  ],
]
`;

exports[`filesByDirectory omits descendant directory paths 1`] = `
Array [
  Array [
    "fixture/level1/level2/level3/file3a",
    "fixture/level1/level2/level3/file3b",
  ],
  Array [
    "fixture/level1/level2/file2a",
    "fixture/level1/level2/file2b",
    "fixture/level1/level2/link-to-directory",
    "fixture/level1/level2/link-to-grand-parent-directory",
    "fixture/level1/level2/link-to-parent-directory",
  ],
]
`;

exports[`filesByDirectory omits descendant directory paths 2`] = `
Array [
  Array [
    "fixture/level1/level2/level3/file3a",
    "fixture/level1/level2/level3/file3b",
  ],
  Array [
    "fixture/level1/level2/file2a",
    "fixture/level1/level2/file2b",
    "fixture/level1/level2/link-to-directory",
    "fixture/level1/level2/link-to-grand-parent-directory",
    "fixture/level1/level2/link-to-parent-directory",
  ],
]
`;

exports[`filesByDirectory omits descendant file paths 1`] = `
Array [
  Array [
    "fixture/level1/level2/level3/file3a",
    "fixture/level1/level2/level3/file3b",
  ],
]
`;

exports[`filesByDirectory omits descendant file paths 2`] = `
Array [
  Array [
    "fixture/level1/level2/level3/file3a",
    "fixture/level1/level2/level3/file3b",
  ],
]
`;

exports[`filesByDirectory omits double directory paths 1`] = `
Array [
  Array [
    "fixture/level1/level2/level3/file3a",
    "fixture/level1/level2/level3/file3b",
  ],
]
`;

exports[`filesByDirectory omits double file paths 1`] = `
Array [
  Array [
    "fixture/level1/file1a",
  ],
]
`;

exports[`filesByDirectory options excludeSymlinks exclude symbolic links when excludeSymlinks is true 1`] = `
Array [
  Array [
    "fixture/level1/level2/level3/file3a",
    "fixture/level1/level2/level3/file3b",
  ],
]
`;

exports[`filesByDirectory options excludeSymlinks exclude symbolic links when excludeSymlinks is true 2`] = `
Array [
  Array [
    "fixture/level1/level2/level3/file3a",
    "fixture/level1/level2/level3/file3b",
  ],
  Array [
    "fixture/level1/level2/file2a",
    "fixture/level1/level2/file2b",
  ],
]
`;

exports[`filesByDirectory options excludeSymlinks exclude symbolic links when excludeSymlinks is true 3`] = `
Array [
  Array [
    "fixture/level1/level2/level3/file3a",
    "fixture/level1/level2/level3/file3b",
  ],
  Array [
    "fixture/level1/level2/file2a",
    "fixture/level1/level2/file2b",
  ],
  Array [
    "fixture/level1/file1a",
    "fixture/level1/file1b",
  ],
]
`;

exports[`filesByDirectory options excludeSymlinks exclude symbolic links when excludeSymlinks is true 4`] = `
Array [
  Array [
    "fixture/level1/level2/level3/file3a",
    "fixture/level1/level2/level3/file3b",
  ],
  Array [
    "fixture/level1/level2/file2a",
    "fixture/level1/level2/file2b",
  ],
]
`;

exports[`filesByDirectory options excludeSymlinks includes symbolic links when excludeSymlinks is false 1`] = `
Array [
  Array [
    "fixture/level1/level2/level3/file3a",
    "fixture/level1/level2/level3/file3b",
  ],
]
`;

exports[`filesByDirectory options excludeSymlinks includes symbolic links when excludeSymlinks is false 2`] = `
Array [
  Array [
    "fixture/level1/level2/level3/file3a",
    "fixture/level1/level2/level3/file3b",
  ],
  Array [
    "fixture/level1/level2/file2a",
    "fixture/level1/level2/file2b",
    "fixture/level1/level2/link-to-directory",
    "fixture/level1/level2/link-to-grand-parent-directory",
    "fixture/level1/level2/link-to-parent-directory",
  ],
]
`;

exports[`filesByDirectory options excludeSymlinks includes symbolic links when excludeSymlinks is false 3`] = `
Array [
  Array [
    "fixture/level1/level2/level3/file3a",
    "fixture/level1/level2/level3/file3b",
  ],
  Array [
    "fixture/level1/level2/file2a",
    "fixture/level1/level2/file2b",
    "fixture/level1/level2/link-to-directory",
    "fixture/level1/level2/link-to-grand-parent-directory",
    "fixture/level1/level2/link-to-parent-directory",
  ],
  Array [
    "fixture/level1/file1a",
    "fixture/level1/file1b",
    "fixture/level1/link-to-descendant-directory",
    "fixture/level1/link-to-descendant-file",
    "fixture/level1/link-to-sibling-directory",
    "fixture/level1/link-to-sibling-file",
    "fixture/level1/link-to-unexisting-file",
  ],
]
`;

exports[`filesByDirectory options excludeSymlinks includes symbolic links when excludeSymlinks is false 4`] = `
Array [
  Array [
    "fixture/level1/level2/level3/file3a",
    "fixture/level1/level2/level3/file3b",
  ],
  Array [
    "fixture/level1/level2/file2a",
    "fixture/level1/level2/file2b",
    "fixture/level1/level2/link-to-directory",
    "fixture/level1/level2/link-to-grand-parent-directory",
    "fixture/level1/level2/link-to-parent-directory",
  ],
]
`;
